datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -579,26 +652,199 @@ where
         )
     }
 
-    fn find_first_blocks_with_logs(
+    fn calls_in_block(
+        &self,
+        logger: &Logger,
+        block_number: u64,
+        block_hash: H256,
+    ) -> Box<Future<Item = Vec<EthereumCall>, Error = Error> + Send> {
+        let eth = self.clone();
+        let addresses = Vec::new();
+        let calls = eth
+            .trace_stream(&logger, block_number, block_number, addresses)
+            .collect()
+            .map(|trace_chunks| match trace_chunks.len() {
+                0 => vec![],
+                _ => trace_chunks.into_iter().flatten().collect(),
+            })
+            .and_then(move |traces| {
+                // `trace_stream` returns all of the traces for the block, and this
+                // includes a trace for the block reward which every block should have.
+                // If there are no traces something has gone wrong.
+                if traces.is_empty() {
+                    return future::err(format_err!(
+                        "Trace stream return no traces for block: number = {}, hash = {}",
+                        block_number,
+                        block_hash,
+                    ));
+                }
+                // Since we can only pull traces by block number and we have
+                // all the traces for the block, we need to ensure that the
+                // block hash for the traces is equal to the desired block hash.
+                for trace in traces.iter() {
```

Comment:
Wouldn't checking the first trace be enough? Even if it's by number, it shouldn't return traces for multiple blocks?

Reply:
Yeah, that _should_ be enough. I updated to reflect this.